### ISSUE TITLE
Expert loss is too high for synchronized samples

### ISSUE THREAD

**Issue by backpass** [Rudrabha/Wav2Lip#132]
Hi, @prajwalkr 

I tried using your expert network to check the "true" loss for ground-truth video/audio pairs (so to understand which value to approximate while training). In different issues and README you are claiming that expert network was  trained to ~0.2 loss, but in my experiments with ground-truth video and audio, I get values between **0.5 - 1.1**.

I tried using different methods to check this, using LRS2 dataset: 
1. sample N random videos and pass arbitrary video/audio pairs to network | got **~0.5-0.6** 
2. using your training code and giving the ground-truth images to get_sync_loss() function during training | got **~0.7-1.1**

So, I am curious to find out why my results contradict yours? And what can I do to correctly reproduce your performance?

Also, I have checked sync-correctness, if it important for LRS2, on small subset of videos and got that many samples have non-zero AV offsets. Does it have something to do with the results above? And did you sync-correct LRS2 dataset?

**Comment by prajwalkr** [comment]
The LRS2 dataset is already sync corrected. Offsets of +/- 1 are okay. 

Please check your preprocessed face images/audios. The released model does give ~0.2 sync loss.

**Comment by backpass** [comment]
Thank you for answer, I will proceed with experimenting.

**Comment by DouYishun** [comment]
@prajwalkr Hi, I also encounter such problems.
I also tried using different methods to check this, using LRS2 dataset, and using released syncnet model:
1. I sample N random videos and pass arbitrary video/audio(chosen = img_name) pairs to network and | got loss ~= 0.5-0.6.
2. I sample N random videos and pass arbitrary wrong video/audio(chosen = wrong_img_name) pairs to network and I got loss ~= 0.2.

The second result proves that the code, env is OK.  (Furthermore, I run inference.py and get nice lip moving.)
And I suspect that an error occurs during the data preprocessing (image/audios).
But when I check the preprocessed samples with high loss, both the cropped face and the audio is OK.

Does there are any other reasons resulting such problems? @prajwalkr 
And how did you solve it? @backpass

**Comment by ahaliassos** [comment]
Hi @prajwalkr,
I also have a similar issue. I downloaded the pretrained weights for the lip sync expert and, as @DouYishun noticed, it seems to almost always predict that the video and audio are misaligned (i.e., the cosine similarity between the embeddings is almost always small), even when the "right" video/audio pairs are passed. This results in ~0.2 loss on misaligned samples but a much higher loss (~0.9 for me) on aligned samples. 

I made sure to preprocess the data using your code (with default parameters) without making any changes that could result in different processing pipelines. Namely, I preprocessed mp4 files (I tried videos from both LRS2 and LRW with similar results) using your preprocessing code (preprocess.py) to extract tight facial crops and the audio. Then, the samples are generated from the Dataset class in color_syncnet_train.py using all default parameters in hparams. I evaluated the pretrained model using the eval_model function in the same file. 

Could it be that some hyperparameters have changed since you evaluated the model, resulting in different preprocessing (such as how the mel spectrogram is generated)? Are you able to obtain ~0.2 loss on both aligned and misaligned data using the current code (and default hyperparameters)? I would really appreciate your help.

Many thanks!

**Comment by prajwalkr** [comment]
The model released is actually very good at detecting aligned samples: > 95% accuracy. Are you trying on LRS2 'main' set videos or test videos? We are not sure if the LRS2 pretrain set is good to test for this sync accuracy.

**Comment by ahaliassos** [comment]
I was actually using the LRS2 pretrain set, but now I just tested on some of the main set videos - I get the same on misaligned samples (~0.2 loss) and ~0.6 loss for aligned samples now (which is better than what I got for the pretrain set but still too high). 

These results are very similar to yours, @DouYishun - did you manage to solve the issue?

**Comment by DouYishun** [comment]
@prajwalkr @ahaliassos  It seems that there are many user trapped in such a dilemma.

**Comment by wangzheliu** [comment]
@DouYishun @ahaliassos @backpass  I also have a similar issue. | got sync_loss ~0.7 during training.loss is not falling.Have you solved the problem.Thank you!

**Comment by zhanghming** [comment]
@prajwalkr , I noticed you mentioned that "The model released is actually very good at detecting aligned samples: > 95% accuracy.", I wonder how to compute the lip sync accuracy in LRS2 val dataset? Are there any codes or clues? I want to evaluate performance of my lip-sync expert quantitatively.
